crates/build/src/buildinfo.rs

```diff
@@ -0,0 +1,288 @@
+use anyhow::Result;
+use std::fmt::Write;
+use std::fs;
+use std::str::FromStr;
+use std::{collections::BTreeMap, fmt, io, iter::FromIterator, path::Path};
+
+/// The default file name used for [BuildInfo] files.
+pub const DEFAULT_FILE_NAME: &'static str = ".build-info";
+/// A collection of variable-value pairs used to reproduce a build of some artifact. These pairs
+/// are representable as either:
+/// - a URI-like string, e.g., `wasmtime?REPOSITORY=https://...+COMMIT=...
+/// - a newline-separated file, e.g.,
+///   ```text
+///   NAME=wasmtime
+///   REPOSITORY=https://...
+///   COMMIT=...
+///   ```
+#[derive(Debug, PartialEq)]
+pub struct BuildInfo(BTreeMap<String, String>); // TODO implement as a `Cow<'a, str>` instead.
+impl BuildInfo {
```

Comment:
Yeah, there are many different lifetimes used. I tried the refactor you suggested just now and immediately ran into more things to fix. I think it is just simpler to leave it as `String` since performance here is not very critical. Removed the TODO.